File: 2021/golang/day02.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func Day02() {
	file, _ := os.Open("../input/day02.input")
	defer file.Close()
	scanner := bufio.NewScanner(file)
	x, y1, y2 := 0, 0, 0

	for scanner.Scan() {
		fields := strings.Fields(scanner.Text())
		value, _ := strconv.Atoi(fields[1])
		switch fields[0] {
		case "forward":
			x += value
			y2 += value * y1
		case "up":
			y1 -= value
		case "down":
			y1 += value
		}
	}

	fmt.Println("Answer A:", x*y1)
	fmt.Println("Answer B:", x*y2)
}

File: 2021/python/day02.py
lines = [i.strip() for i in open("../input/day02.input", encoding="utf-8").readlines()]
D = {"forward": 1j, "down": 1, "up": -1}

pos = sum([D[line.split()[0]] * int(line.split()[1]) for line in lines])
print("Answer A:", int(pos.imag * pos.real))

x, y, aim = 0, 0, 0
for direction, steps in [line.split() for line in lines]:
    if direction == "forward":
        x += int(steps)
        y += aim * int(steps)
    else:
        aim += int(steps) * D[direction]
print("Answer B:", x * y)
